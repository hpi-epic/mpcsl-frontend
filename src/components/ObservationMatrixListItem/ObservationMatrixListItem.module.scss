.ObservationMatrixListItem {
  max-width: 300px;
  min-width: 300px;
  margin: 5px;
  flex-grow: 1;
}

.ObservationMatrixListItemContent {
  height: 250px;
  display: flex;
  flex-direction: column;
  p {
    flex-grow: 1;
  }
}

.Description {
  height: 152px;
  overflow: hidden;
}

.ListButton {
  // margin-right: 5px;
  margin: 5px;
}
