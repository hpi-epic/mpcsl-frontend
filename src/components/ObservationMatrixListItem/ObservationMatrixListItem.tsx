import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Card, Progress, Button, message, Popconfirm } from 'antd';
import Axios, { AxiosRequestConfig } from 'axios';
import styles from './ObservationMatrixListItem.module.scss';
import { deleteObservationMatrix } from '../../actions/apiRequests';

interface IObservationMatrixListElement {
  id: number;
  load_query: string;
  name: string;
  description?: string;
  data_source?: string;
  time_created?: string;
  onClick: () => void;
}

const ObservationMatrixListItem = (props: IObservationMatrixListElement) => {
  const { name, description } = props;
  const [inputRef, setInputRef] = useState<HTMLInputElement>();
  const [uploadProgress, setUploadProgress] = useState<number | undefined>();
  const history = useHistory();
  return (
    <Card
      title={name}
      hoverable
      className={styles.ObservationMatrixListItem}
      onClick={() => history.push(`/${props.id}/experiments`)}
    >
      <div className={styles.ObservationMatrixListItemContent}>
        <p className={styles.Description}>{description}</p>
        <div style={{ alignSelf: 'flex-end' }}>
          <div style={{ width: 170 }}>
            {uploadProgress ? (
              <Progress percent={uploadProgress} size="small" />
            ) : null}
          </div>
          <Button
            className={styles.ListButton}
            type="primary"
            ghost={true}
            onClick={e => {
              e.stopPropagation();
              if (inputRef) {
                inputRef.click();
              }
            }}
          >
            Upload Ground Truth
          </Button>
          <input
            style={{ display: 'none' }}
            type="file"
            name="file"
            id="file"
            ref={ref => (ref ? setInputRef(ref) : undefined)}
            accept=".gml"
            onClick={e => e.stopPropagation()}
            onChange={e => {
              const target = e.target;
              const file =
                target.files && target.files.length === 1
                  ? target.files[0]
                  : undefined;
              if (file) {
                const data = new FormData();
                data.append('graph_file', file);

                const config: AxiosRequestConfig = {
                  onUploadProgress: progressEvent =>
                    setUploadProgress(
                      Math.round(
                        (progressEvent.loaded * 100) / progressEvent.total
                      )
                    )
                };
                Axios.post(`/api/dataset/${props.id}/upload`, data, config)
                  .catch(err => {
                    message.error(err.response.data.message);
                  })
                  .finally(() => {
                    setTimeout(() => setUploadProgress(undefined), 1000);
                    target.value = '';
                  });
              }
            }}
          />
          <Button
            className={styles.ListButton}
            onClick={e => {
              e.stopPropagation();
              props.onClick();
            }}
            type="primary"
            ghost={true}
          >
            View
          </Button>
          <Popconfirm
            title="Delete Matrix?"
            onConfirm={e => {
              e?.stopPropagation();
              deleteObservationMatrix(props);
            }}
            onCancel={e => e?.stopPropagation()}
          >
            <Button
              icon="delete"
              type="danger"
              className={styles.ListButton}
              ghost={true}
              onClick={e => e.stopPropagation()}
            >
              Delete
            </Button>
          </Popconfirm>
        </div>
      </div>
    </Card>
  );
};

export { ObservationMatrixListItem };
