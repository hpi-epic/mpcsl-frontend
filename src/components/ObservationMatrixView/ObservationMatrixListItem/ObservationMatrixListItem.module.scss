.ObservationMatrixListItem {
  max-width: 300px;
  min-width: 300px;
  margin: 5px;
  flex-grow: 1;
}

.ObservationMatrixListItemContent {
  height: 200px;
  display: flex;
  flex-direction: column;

  p {
    flex-grow: 1;
  }
}

.ListButton {
  // margin-right: 5px;
  margin: 5px;
}

.Description {
  width: 100%;
}
