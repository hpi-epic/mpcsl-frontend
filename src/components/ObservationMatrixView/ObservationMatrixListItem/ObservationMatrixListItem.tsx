import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Card, Descriptions, message, Modal, Progress, Tooltip } from 'antd';
import Axios, { AxiosRequestConfig } from 'axios';
import styles from './ObservationMatrixListItem.module.scss';
import {
  deleteObservationMatrix,
  getObservationMatrixMetadata
} from '../../../restAPI/apiRequests';
import { IObservationMatrixMetadata } from '../../../types/types';
import {
  CheckCircleTwoTone,
  DeleteOutlined,
  InfoCircleOutlined,
  UploadOutlined
} from '@ant-design/icons';

const { confirm } = Modal;

interface IObservationMatrixListElement {
  id: number;
  load_query: string;
  name: string;
  description?: string;
  data_source?: string;
  time_created?: string;
  onClick: () => void;
}

const ObservationMatrixListItem: React.FC<IObservationMatrixListElement> = ({
  id,
  load_query,
  name,
  description,
  data_source,
  time_created,
  onClick
}) => {
  const [inputRef, setInputRef] = useState<HTMLInputElement>();
  const [uploadProgress, setUploadProgress] = useState<number | undefined>();
  const [metadata, setMetadata] = useState<
    IObservationMatrixMetadata | undefined
  >();
  useEffect(() => {
    getObservationMatrixMetadata(id).then(setMetadata);
  }, [id, load_query, name, description, data_source, time_created, onClick]);
  const history = useHistory();
  return (
    <Card
      title={
        <div
          style={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'space-between'
          }}
        >
          {name}
          {metadata?.has_ground_truth ? (
            <Tooltip title="Observation Matrix has Ground Truth Graph">
              <CheckCircleTwoTone
                style={{ fontSize: 20 }}
                type="check-circle"
                twoToneColor="#52c41a"
              />
            </Tooltip>
          ) : null}
        </div>
      }
      hoverable
      className={styles.ObservationMatrixListItem}
      onClick={() => history.push(`/${id}/experiments`)}
      actions={[
        <Tooltip key="upload" title="Upload Ground Truth Graph">
          <UploadOutlined
            style={{ fontSize: 20 }}
            onClick={e => {
              e.stopPropagation();
              if (inputRef) {
                inputRef.click();
              }
            }}
          />
        </Tooltip>,
        <Tooltip key="details" title="Show Details">
          <InfoCircleOutlined
            style={{ fontSize: 20 }}
            onClick={e => {
              e.stopPropagation();
              onClick();
            }}
          />
        </Tooltip>,
        <Tooltip key="delete" title="Delete Observation Matrix">
          <DeleteOutlined
            style={{ fontSize: 20 }}
            onClick={e => {
              e?.stopPropagation();
              confirm({
                title:
                  'Do you want to delete the following Observation Matrix?',
                content: `${name} - ${description}`,
                onOk() {
                  deleteObservationMatrix(id, name).catch();
                }
              });
            }}
          />
        </Tooltip>
      ]}
    >
      <div className={styles.ObservationMatrixListItemContent}>
        <Descriptions size="small" layout="vertical" column={1}>
          <Descriptions.Item className={styles.Description} label="Description">
            <p
              style={{
                marginTop: -10,
                height: 64,
                overflow: 'hidden',
                width: 250,
                color: 'rgba(0,0,0,.65)',
                overflowWrap: 'break-word'
              }}
            >
              {description}
            </p>
          </Descriptions.Item>
        </Descriptions>
        <Descriptions size="small" column={1}>
          {metadata ? (
            <>
              <Descriptions.Item label="Variables">
                {metadata.variables}
              </Descriptions.Item>
              <Descriptions.Item label="Observations">
                {metadata.observations}
              </Descriptions.Item>
              <Descriptions.Item label="Time created">
                {new Date(metadata.time_created * 1000).toLocaleString('de-DE')}
              </Descriptions.Item>
            </>
          ) : null}
        </Descriptions>
        <div style={{ alignSelf: 'flex-end' }}>
          <div style={{ width: 170 }}>
            {uploadProgress ? (
              <Progress percent={uploadProgress} size="small" />
            ) : null}
          </div>

          <input
            style={{ display: 'none' }}
            type="file"
            name="file"
            id="file"
            ref={ref => (ref ? setInputRef(ref) : undefined)}
            accept=".gml"
            onClick={e => e.stopPropagation()}
            onChange={e => {
              const target = e.target;
              const file =
                target.files && target.files.length === 1
                  ? target.files[0]
                  : undefined;
              if (file) {
                const data = new FormData();
                data.append('graph_file', file);

                const config: AxiosRequestConfig = {
                  onUploadProgress: progressEvent =>
                    setUploadProgress(
                      Math.round(
                        (progressEvent.loaded * 100) / progressEvent.total
                      )
                    )
                };
                Axios.post(`/api/dataset/${id}/upload`, data, config)
                  .catch(err => {
                    message.error(err.response.data.message);
                  })
                  .finally(() => {
                    setTimeout(() => setUploadProgress(undefined), 1000);
                    target.value = '';
                  });
              }
            }}
          />
        </div>
      </div>
    </Card>
  );
};

export { ObservationMatrixListItem };
