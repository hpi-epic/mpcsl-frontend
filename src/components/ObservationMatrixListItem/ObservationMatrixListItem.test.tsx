import React from 'react';
import renderer from 'react-test-renderer';
import { ObservationMatrixListItem } from './ObservationMatrixListItem';
import { MemoryRouter } from 'react-router-dom';

describe('<ListElementExperiment />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <ObservationMatrixListItem
            id={0}
            load_query="SELECT * FROM DUMMY"
            name="test"
            onClick={() => undefined}
          />
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
