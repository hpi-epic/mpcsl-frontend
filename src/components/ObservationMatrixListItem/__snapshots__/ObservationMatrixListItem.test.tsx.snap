// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListElementExperiment /> renders correctly 1`] = `
<div
  className="ant-card ObservationMatrixListItem ant-card-bordered ant-card-hoverable"
  onClick={[Function]}
>
  <div
    className="ant-card-head"
    style={Object {}}
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        test
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="ObservationMatrixListItemContent"
    >
      <p
        className="Description"
      />
      <div
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
      >
        <div
          style={
            Object {
              "width": 170,
            }
          }
        />
        <button
          className="ant-btn ListButton ant-btn-primary ant-btn-background-ghost"
          onClick={[Function]}
          type="button"
        >
          <span>
            Upload Ground Truth Graph
          </span>
        </button>
        <input
          accept=".gml"
          id="file"
          name="file"
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
        />
        <button
          className="ant-btn ListButton ant-btn-primary ant-btn-background-ghost"
          onClick={[Function]}
          type="button"
        >
          <span>
            View
          </span>
        </button>
        <button
          className="ant-btn ListButton ant-btn-danger ant-btn-background-ghost"
          onClick={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          type="button"
        >
          <i
            aria-label="icon: delete"
            className="anticon anticon-delete"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="delete"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
              />
            </svg>
          </i>
          <span>
            Delete
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
